File: EAS520/Assignments/Assignment_1/r_code/p5_mv_prod.R
library(pracma)

n <- 10000
A <- rand(n, n)
b <- runif(n)

C <- zeros(n, 1)

# Using vectorization
t1 <- Sys.time()
CC = A%*%b
t2 <- Sys.time()

time_vec = t2 - t1


# Using for-loop

t3 <- Sys.time()
for (i in 1:n) {
  for (j in 1:n) {
    C[i] <- C[i] + A[i, j]*b[j]
  }
}
t4 <- Sys.time()

time_loop = t4 - t3


norm <- norm(CC-C)

speed_up <- as.double(time_loop, units='secs')/as.double(time_vec, units='secs')

n
norm
speed_up

File: EAS520/Assignments/Assignment_1/matlab_code/p5_mv_prod.m
clear all;

n = 100; 

A = rand(n,n); % Matrix
x = rand(n,1); % vector

b = zeros(n,1); 
bb = zeros(n,1);

% For-loop
tic
for i = 1:n
    for j=1:n
    b(i) = b(i) + A(i,j)*x(j);
    end
end
timeloop = toc;

% Loop Vectorization
tic
for i = 1:n
    bb(i) = A(i,:)*x;
end
timeloopvec = toc;

% vectorization
tic
    bbb = A*x;
timevec = toc;

fprintf("norm 1: %f", norm(b-bb))
fprintf("\nnorm 2: %f\n", norm(b-bbb))

Speedup = timeloop/timeloopvec;
Speedup2 = timeloop/timevec;
Speedup3 = timeloopvec/timevec;

fprintf("\nSpeedup: %f\n", Speedup);
fprintf("\nSpeedup 2: %f\n", Speedup2);
fprintf("\nSpeedup 3: %f\n", Speedup3);
